### ISSUE TITLE
Failure to find package with period in name

### ISSUE THREAD

**Issue by WillDaSilva** [steiza/simplepypi#15]
Some tools (e.g. Flit) produce packages with a period in their name (e.g. `abc.xyz`) when building a [native namespace package](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages).

When build with pip + Flit the resulting wheel file has the periods replaced by an underscore (e.g. `abc_xyz-0.1.2-py3-none-any.whl`). This wheel can be uploaded to a simplepypi server, but cannot be downloaded.

When I check `http://localhost:8000/packages` (where the simplepypi server is located) the package appears in the list as `abc.xyz`, and when clicked it brings me to `http://localhost:8000/packages/abc.xyz`, which shows the wheel file `abc_xyz-0.1.2-py3-none-any.whl`.

I have tried to download it using `abc.xyz`:
```
$ pip install -i http://localhost:8000/pypi 'abc.xyz'
Looking in indexes: http://localhost:8000/pypi
ERROR: Could not find a version that satisfies the requirement abc.xyz (from versions: none)
ERROR: No matching distribution found for abc.xyz
```

And with `abc_xyz`:
```
$ pip install -i http://localhost:8000/pypi 'abc_xyz'
Looking in indexes: http://localhost:8000/pypi
ERROR: Could not find a version that satisfies the requirement abc_xyz (from versions: none)
ERROR: No matching distribution found for abc_xyz
```

I am able to download other packages from `http://localhost:8000/pypi`  so long as they do not have a period in their name.

I am able to download packages with periods in their name from the official PyPI server at `pypi.org`, and from a PyPI server hosted by [Nexus](https://www.sonatype.com/products/repository-pro).

**Comment by steiza** [comment]
Oh interesting, I haven't worked with packages with a period in their name before. I'm open to changing the implementation, but it isn't clear to me yet what the standard is here.

I made a test package (using `python -m build`, instead of Flit), which happily produced `abc.xyz-0.0.1-py3-none-any.whl` that I uploaded to simplepypi with `twine upload --repository-url http://127.0.0.1:8000 dist/abc.xyz-0.0.1-py3-none-any.whl`.

I was able to view and download that package with the web interface (here's the server logs):
```
INFO - 200 GET /packages/abc.xyz (127.0.0.1) 0.72ms
INFO - 200 GET /packages/abc.xyz/0.0.1/abc.xyz-0.0.1-py3-none-any.whl (127.0.0.1) 1.01ms
```

... but when I did `pip install -i http://127.0.0.1:8000/pypi abc.xyz` that did not work (again, here are the server logs):

```
WARNING - 404 GET /pypi/abc-xyz/ (127.0.0.1) 0.57ms
```

**So it looks like `pip` is substituting `.` for `-` when it queries the server**. But it sounds like your pip + Flit wheel replaces `.` with `_`?

It seems like simplepypi should follow pip's lead and replace `.` with `-`, but I'm not sure that will solve your problem - is there additional information you could provide about how you ended up with `.` replaced with `_`? Like what's your project name in `pyproject.toml`? I don't think there is, but just in case, is there a `setup.cfg` / `setup.py` with a name specified as well?

**Comment by WillDaSilva** [comment]
Thank you for looking into this.

tl;dr: I believe the `normalize` function (below) should be applied to package names when they are being uploaded to the simplepypi server.

```python
import re

def normalize(name: str) -> str:
    return re.sub(r'[-_.]+', '-', name).lower()
```

-----------------

I do not have a `setup.cfg` or `setup.py` for this project. The name in `pyproject.toml` is `abc.xyz`.

Some justification for the name choice: it was chosen to be this since that's the only way to use Flit's namespace package feature. `abc.xyz` is installed such that `from abc import xyz` is possible - this allows for improved organization within the company I'm working for, as we can keep the `xyz` part in a separate repository from other parts of the `abc` package.

It looks like the standard for Python package name normalization is documented in [PEP503](https://www.python.org/dev/peps/pep-0503/#normalized-names).

I captured the HTTPS requests for `pip install [flags omitted] abc.xyz`, which downloaded the wheel file off of my company's PyPI server (hosted via Nexus):

```
https://nexus.[A].com/[B]/simple/abc-xyz
https://nexus.[A].com/[B]/packages/abc-xyz/0.1.2/abc_xyz-0.1.2-py3-none-any.whl
```

The parts inside the square brackets have been omitted. I assume these are `GET` requests.

It looks like my request for the package `abc.xyz` is being normalized into `abc-xyz` by pip, which is then requested from the PyPI server. The PyPI server normalized the names of the packages it stores when they were initially uploaded. When I open `https://nexus.[A].com/[B]/simple/` in a web browser I see that the period has been replaced by a hyphen in the package index.

It looks like the normalization does not apply to the wheel file when it is uploaded, but that's not necessary since by the time the wheel file itself is being selected, the package name and version have already been dealt with (because we are already in the `<package name>/<version number>/` subdirectory.

**Comment by steiza** [comment]
Makes sense to me! I believe this should be fixed with https://github.com/steiza/simplepypi/commit/f0906b6c4880b8481c72d62c41cd94595d900091
